tesseract_environment/src/core/environment.cpp

```diff
@@ -1061,12 +1067,51 @@ bool Environment::addSceneGraph(const tesseract_scene_graph::SceneGraph& scene_g
 
 bool Environment::applyAddCommand(AddLinkCommand::ConstPtr cmd)
 {
-  if (!cmd->getLink() && !cmd->getJoint())
+  assert(!(!cmd->getLink() && !cmd->getJoint()));
+
+  bool link_exists = false;
+  bool joint_exists = false;
+  std::string link_name, joint_name;
+
+  if (cmd->getLink() != nullptr)
+  {
+    link_name = cmd->getLink()->getName();
+    link_exists = (scene_graph_->getLink(link_name) != nullptr);
+  }
+
+  if (cmd->getJoint() != nullptr)
+  {
+    joint_name = cmd->getJoint()->getName();
+    joint_exists = (scene_graph_->getJoint(joint_name) != nullptr);
+  }
+
+  if ((link_exists && !cmd->replaceAllowed()))
+  {
+    CONSOLE_BRIDGE_logWarn("Tried to add link (%s) which already exists. Set replace_allowed to enable replacing.",
+                           link_name.c_str());
     return false;
+  }
 
-  std::string link_name = cmd->getLink()->getName();
-  if (!cmd->getJoint())
+  if ((link_exists && cmd->getJoint()))
   {
+    CONSOLE_BRIDGE_logWarn("Tried to add link (%s) which already exists with a joint provided. This is not supported.",
+                           link_name.c_str());
+    return false;
+  }
```

Comment:
Why isn't replaceing the link and joint at the same time supported? Can't we just call the same replaceJoint function in the scene graph?

Reply:
See chart below